File: 133. Clone Graph.py
# https://leetcode.com/problems/clone-graph/

# Given a reference of a node in a connected undirected graph.

# Return a deep copy (clone) of the graph.

# Each node in the graph contains a val (int) and a list (List[Node]) of its
# neighbors.

###############################################################################

# bfs (deque)
# use node_copy_map{key: value} with key = node and val = node.copy

"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node: return node
        
        node_copy_map = {} # key = node, val = node_copy
        node_copy_map[node] = Node(node.val)
        deque = collections.deque([node])
        
        while deque:
            curr = deque.popleft()
            for neigh in curr.neighbors: # loop over all neighbors
                if neigh not in node_copy_map: # new node, copy node.val
                    node_copy_map[neigh] = Node(neigh.val)
                    deque.append(neigh)
                node_copy_map[curr].neighbors.append(node_copy_map[neigh]) # copy node.neighbors
        
        return node_copy_map[node] # return copy of fisrt node

File: 133. Clone Graph.cpp
// https://leetcode.com/problems/clone-graph/

// Given a reference of a node in a connected undirected graph.

// Return a deep copy (clone) of the graph.

// Each node in the graph contains a value (int) and a list (List[Node]) of its
// neighbors.

////////////////////////////////////////////////////////////////////////////////

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (node == nullptr) return node;
        unordered_map<Node*, Node*> copies; // copies[old node] = copied node
        copies[node] = new Node(node->val);
        
        queue<Node*> todo; todo.push(node);
        while (!todo.empty()) {
            Node* curr = todo.front(); todo.pop();
            for (Node* neigh : curr->neighbors) {
                if (copies.find(neigh) == copies.end()) {
                    copies[neigh] = new Node(neigh->val);
                    todo.push(neigh);
                }
                copies[curr]->neighbors.push_back(copies[neigh]);
            }
        }
        return copies[node];
    }
};
